text/subsingleton-elimination.md

```diff
@@ -0,0 +1,176 @@
+
+The status of casts in the implementation is however fragile, so this may require some clarification of casts first.
+
+## New command to support arbitrary `HProp`
+
+We propose a new command:
+```
+Subsingleton id context : ishProp A := proof.
+```
+which recast the type of `A` from `context -> Type` to `context -> HProp` so that it is latter considered impredicative.
+
+An alternative is to have a modifier to definitions in the form:
+```
+Subsingleton id context := def : HProp by proof.
+```
+But then, we probably want also:
+```
+Subsingleton id context := def : Hprop.
+Proof.
+the proof of being hProp
+Qed.
+```
+which may be complicated.
+
+If `HProp` is made distinct from `Prop`, new declarations `Inductive ... : ... -> HProp` would be accepted too. When to tell that the type has subsingleton elimination? There is a proposal in the next section.
+
+## Dynamic detection of subsingleton elimination
+
+We propose an extension of `match` of the following form:
+```
+match t as id in I params return typ by u with
+| C1 vars => body
+...
+| Cn vars => body
+end
+```
+where, when `I` is `HProp`-truncated but the instance `I args` is an `HProp` proved by `u : forall x y:I args, x = y`, the elimination to `Type` is allowed.
+
+# Side remark on template polymorphism for `Prop`
+
+Instances of parametric types which are not in `Prop` in the general case are automatically recognized as belonging to the `Prop` class (and treated as subsingleton for `match` and as impredicative in function types) when the parameters are themselves in `Prop`. This is typically the case of `prod A B` which behaves as `and A B` when `A` and `B` are in `Prop`.
+
+Template polymorphism induces a view where an inductive type in `Prop` does not necessarily mean explicitly `Prop`-truncated. It is so only when the original type is not a syntactic `Prop`-subsingleton.
+
+Without template polymorphism, explicit coercions from `prod A B` to `and A B` should be inserted (when `A` and `B` are `Prop`), and explicit Prop-boxing of `prod A B` should be done to make `prod A B` behave impredicatively.
+
+# Issues with extraction of `HProp`
+
+Types in `HProp` which are not in the current syntactic class `Prop` could simply be extracted as if they were types.
+
+# Conclusion
+
+A few extensions could be done pretty easily:
+1. accept singleton types with all arguments in `SProp` to be in `SProp` (`True`, `and`, ...)
+2. accept in `SProp` and `Prop` singleton types with arguments of the form `forall a:A, B` when `B` is recursively an `SProp` or `Prop`-subsingleton
+4. dynamically recognize subsingleton elimination for `Prop` and `SProp` in `match`
+
+The "disjoint indices" extension would require some days of work:
+
+4. accept in `SProp` and `Prop` types with several constructors when (small) indices justify that the constructors are disjoint (section 5 of "Definitional Proof-Irrelevance without K")
```

Comment:
I see this in 2 parts:
- having indices at all (in the non 0 constructor case), in which case the implicit equality must be small (except for prop without -indices-matter)
  this was actually already present in an early version of SProp, see https://github.com/coq/ceps/pull/37/files#diff-dce84f9d9bcdf12e5ebabb0ce1c8791176850fe3aa48a8e58f068d016c92a90dR255 https://github.com/coq/coq/issues/7929 and https://github.com/coq/coq/blob/089f1504293559b4472f8f4f13c2a498a51a5d64/kernel/indtypes.ml#L116-L222 and related code in that branch
- once we have indices, allowing multiple constructors when the indices force them to be disjoint

Reply:
> this was actually already present in an early version of SProp

And you removed it because it raised issues?

> (except for prop without -indices-matter)

It would be nice to eventually have `-indices-matter` the default.

> once we have indices, allowing multiple constructors when the indices force them to be disjoint

OK, thus there are plans?!

Incidentally, at some time, I tried using `SProp` to have irrelevant proofs of `le` but this was difficult (I should report about my experience; maybe I just don't know the good tricks...).

Reply:
>And you removed it because it raised issues?

It was somewhat incomplete (for instance extraction didn't handle it) and didn't seem all that useful.

>OK, thus there are plans?!

It's just "if we do it, it will probably be something like this"